ariadne_codegen/plugins/shorter_results.py

```diff
@@ -0,0 +1,294 @@
+                    value=generate_attribute(
+                        value=previous_yield_value,
+                        attr=single_field_return_class,
+                    )
+                )
+            ),
+        )
+
+        self._update_imports(method_def, single_field_class)
+
+        return super().generate_client_method(method_def)
+
+    def _generate_sync_client_method(
+        self,
+        method_def: Union[ast.FunctionDef, ast.AsyncFunctionDef],
+        return_stmt: ast.Return,
+    ):
+        if return_stmt.value is None:
+            return super().generate_client_method(method_def)
+
+        if not isinstance(method_def.returns, ast.Name):
+            return super().generate_client_method(method_def)
+
+        node_and_class = _return_or_yield_node_and_class(
+            method_def.returns.id, self.class_dict
+        )
+        if node_and_class is None:
+            return super().generate_client_method(method_def)
+
+        return_node, single_field_class, single_field_return_class = node_and_class
+
+        # Update the return type to be the inner field and ensure we reference
+        # the single field before returning.
+        method_def.returns = return_node
+        method_def.body[-1] = generate_return(
+            value=generate_attribute(
+                value=return_stmt.value, attr=single_field_return_class
+            )
+        )
+
+        self._update_imports(method_def, single_field_class)
+
+        return super().generate_client_method(method_def)
+
+    def _update_imports(self, method_def, single_field_class):
+        if single_field_class not in self.class_dict:
+            return
+
+        # After we change the type we also need to import it in the client if
+        # it's one of our generated types so add the extra import as needed.
+        if method_def.name not in self.extended_imports:
+            self.extended_imports[method_def.name] = set()
+
+        self.extended_imports[method_def.name].add(single_field_class)
+
+
+def _return_or_yield_node_and_class(
+    current_return_class: str, class_dict: Dict[str, ast.ClassDef]
+) -> Optional[tuple[ast.expr, str, str]]:
+    if current_return_class not in class_dict:
+        return None
+
+    return_class_ast = class_dict[current_return_class]
+
+    # If we're not a single method class there's nothing for the plugin to do.
+    if len(return_class_ast.body) != 1:
+        return None
+
+    # If we are a single method class, ensure we inherit from BaseModel. If we
+    # inherit from another class like a fragment that could contain more fields
+    # so we don't want to unpack.
+    if len(return_class_ast.bases) != 1:
+        return None
+
+    if not isinstance(return_class_ast.bases[0], ast.Name):
+        return None
+
+    if not return_class_ast.bases[0].id == "BaseModel":
+        return None
+
+    single_field = return_class_ast.body[0]
+    if single_field is None:
+        return None
```

Comment:
I think we should "unpack" fragments, for example:
```py
def _get_all_fields(
    class_def: ast.ClassDef, class_dict: Dict[str, ast.ClassDef]
) -> List[ast.AnnAssign]:
    fields = []
    for base in class_def.bases:
        if not isinstance(base, ast.Name):
            continue
        fields.extend(_get_all_fields(class_dict[base.id], class_dict=class_dict))

    for field in class_def.body:
        if isinstance(field, ast.AnnAssign):
            fields.append(field)

    return fields
```
And then:
```suggestion
    fields = _get_all_fields(class_dict[current_return_class], class_dict)
    if not fields or len(fields)!= 0:
        return None
    single_field = fields[0]
```

I see that for now this might be difficult, as `generate_result_class` for fragments is called after `generate_client_method` but I just created pr to fix this.

Reply:
Awesome suggestion, makes sense! Added this but there are still some issues.

1. We return if we don't have the class in our class dict. This is what we want for `BaseClass` f.ex. but is there any risk that this could be a class with actual fields?
2. Even with this I still don't generate fragments first. It's _called_ by the fragments, but that's after I generated the method. I tried adding this piece:

```python
    def generate_fragments_module(
        self,
        module: ast.Module,
        fragments_definitions: Dict[str, FragmentDefinitionNode],
    ) -> ast.Module:
        class_defs = list(filter(lambda x: isinstance(x, ast.ClassDef), module.body))
        for class_def in class_defs:
            if not isinstance(class_def, ast.ClassDef):
                continue

            self.class_dict[class_def.name] = class_def

        return module
```

But that gets called _after_ `generate_client_method` still. 🤔

Reply:
1. Generated model can only inherit from `BaseModel` (which we can safely assume doesn't have any fields we should care about) or fragments mixins. Each fragment mixin can only inherit from other fragments or `BaseModel`. I see that in my  `_get_all_fields` I didn't handle the fact that `BaseModel` is not in `class_dict`.
```py
def _get_all_fields(
    class_def: ast.ClassDef, class_dict: Dict[str, ast.ClassDef]
) -> List[ast.AnnAssign]:
    fields = []
    for base in class_def.bases:
        if not isinstance(base, ast.Name) or base.id not in class_dict:
            continue
        fields.extend(_get_all_fields(class_dict[base.id], class_dict=class_dict))

    for field in class_def.body:
        if isinstance(field, ast.AnnAssign):
            fields.append(field)

    return fields
```
2. Oh, you are right. But now we can modify methods in `generate_client_module` which is called after all `generate_result_class`, including models for fragments. Then we don't need `generate_fragments_module`, and it could look like this:
```py
    def generate_client_module(self, module: ast.Module) -> ast.Module:
        client_def = next(
            filter(lambda o: isinstance(o, ast.ClassDef), module.body), None
        )
        if not client_def:
            return super().generate_client_module(module)

        for method_def in [
            m
            for m in client_def.body
            if isinstance(m, (ast.FunctionDef, ast.AsyncFunctionDef))
        ]:
            self._modify_method_def(method_def)

        # imports stuff which is already there
        return super().generate_client_module(module)
    
    def _modift_method_def(self, method_def: Union[ast.FunctionDef, ast.AsyncFunctionDef]):
        # stuff which currently is in generate_client_method
```
This would also require changing how current validation works, so 1 "invalid" method wouldn't stop us from updating the rest of them. What do you think about this approach?

Reply:
EDIT: I think we wrote a response at almost the same time 😅  Plugin's instance is created once at the beginning of  script execution

Reply:
Figured out why. If we don't specify any fields in `TestQuery` that class will be empty like you wrote before:

```python
class TestQuery(TestFragment):
    pass
```

When a class is empty it will never be passed to `generate_result_class` so we won't add it to our dict and thus we can't see what class it extends. I'll see if I can figure out how we can still pass the class just for plugin knowledge.

Reply:
Oh, that shouldn't be the case. I'll look into this, maybe I broke this when I pushed changes related to fragments.

Reply:
Looks to be related to `ariadne_codegen/client_generators/result_types.py` and `_parse_type_definition`. We only call `generate_result_class` when iterating over `resolved_selection_set`. One solution to fix this would be to call the plugin class if the selection set is empty:

```diff
diff --git a/ariadne_codegen/client_generators/result_types.py b/ariadne_codegen/client_generators/result_types.py
index a3f895d..9bca05e 100644
--- a/ariadne_codegen/client_generators/result_types.py
+++ b/ariadne_codegen/client_generators/result_types.py
@@ -256,6 +256,13 @@ class ResultTypesGenerator:
                     selection_set=selection_set,
                 )
 
+        if len(resolved_selection_set) == 0 and self.plugin_manager:
+            class_def = self.plugin_manager.generate_result_class(
+                class_def,
+                operation_definition=self.operation_definition,
+                selection_set=selection_set,
+            )
+
         if not class_def.body:
             class_def.body.append(generate_pass())
 ```

But maybe that's a bad approach.

Reply:
Yeah, I found the same issue, turns out it was broken even before fragments change 😓 I created [pr](https://github.com/mirumee/ariadne-codegen/pull/147) to fix this